File: runtests/python_lua/lua_construct.py
from __future__ import annotations as __01asda1ha
from lark import Token as token
import dataclasses as dataclasses
import typing as typing
from .lua_require import (mkBinOpSeq,mkOperand,mkOperator,op,none,some,maybe,listMap,appendList)

@dataclasses.dataclass
class DotName:
    value0: funcname
    value1: token

@dataclasses.dataclass
class MethodName:
    value0: funcname
    value1: token

@dataclasses.dataclass
class VarName:
    value: token

if typing.TYPE_CHECKING:
    funcname = typing.Union[VarName,MethodName,DotName]
else:
    funcname = (VarName,MethodName,DotName)

@dataclasses.dataclass
class PositionalArgs:
    value0: token
    value1: list[expr]

@dataclasses.dataclass
class StringArg:
    value: token

@dataclasses.dataclass
class TableArgs:
    value: table

if typing.TYPE_CHECKING:
    arguments = typing.Union[TableArgs,StringArg,PositionalArgs]
else:
    arguments = (TableArgs,StringArg,PositionalArgs)

@dataclasses.dataclass
class TableConstructor:
    value0: token
    value1: list[table_field]

if typing.TYPE_CHECKING:
    table = TableConstructor
else:
    table = (TableConstructor)

@dataclasses.dataclass
class ElementField:
    value0: expr

@dataclasses.dataclass
class IndexField:
    pos: token
    value: expr
    item: expr

@dataclasses.dataclass
class NameField:
    key: token
    value: expr

if typing.TYPE_CHECKING:
    table_field = typing.Union[NameField,IndexField,ElementField]
else:
    table_field = (NameField,IndexField,ElementField)

@dataclasses.dataclass
class Assignment:
    value0: bool
    value1: list[expr]
    value2: maybe[list[expr]]

@dataclasses.dataclass
class BreakStmt:
    value: token

@dataclasses.dataclass
class DoStmt:
    value0: token
    value1: block

@dataclasses.dataclass
class EmptyStmt:
    value0: token

@dataclasses.dataclass
class ExprStmt:
    value: expr

@dataclasses.dataclass
class ForInStmt:
    value0: token
    value1: list[token]
    value2: list[expr]
    value3: block

@dataclasses.dataclass
class ForRangeStmt:
    value0: token
    value1: token
    value2: range
    value3: block

@dataclasses.dataclass
class GotoStmt:
    value0: token
    value1: token

@dataclasses.dataclass
class IfStmt:
    value0: token
    value1: expr
    value2: block
    value3: list[if_elseif]
    value4: maybe[if_else]

@dataclasses.dataclass
class LabelStmt:
    value: token

@dataclasses.dataclass
class RepeatStmt:
    value0: token
    value1: block
    value2: expr

@dataclasses.dataclass
class ReturnStmt:
    value0: token
    value1: list[expr]

@dataclasses.dataclass
class WhileStmt:
    value0: token
    value1: expr
    value2: block

if typing.TYPE_CHECKING:
    stmt = typing.Union[WhileStmt,ReturnStmt,RepeatStmt,LabelStmt,IfStmt,GotoStmt,ForRangeStmt,ForInStmt,ExprStmt,EmptyStmt,DoStmt,BreakStmt,Assignment]
else:
    stmt = (WhileStmt,ReturnStmt,RepeatStmt,LabelStmt,IfStmt,GotoStmt,ForRangeStmt,ForInStmt,ExprStmt,EmptyStmt,DoStmt,BreakStmt,Assignment)

@dataclasses.dataclass
class Attr:
    value0: expr
    value1: token

@dataclasses.dataclass
class Bin:
    op: token
    l: expr
    r: expr

@dataclasses.dataclass
class Bool:
    value0: token
    value1: bool

@dataclasses.dataclass
class CallFunc:
    value0: expr
    value1: arguments

@dataclasses.dataclass
class CallMethod:
    value0: expr
    value1: token
    value2: arguments

@dataclasses.dataclass
class Ellipse:
    value0: token

@dataclasses.dataclass
class Exponent:
    value0: expr
    value1: expr

@dataclasses.dataclass
class FuncDef:
    pos: token
    is_local: bool
    fname: maybe[funcname]
    params: maybe[params]
    body: block

@dataclasses.dataclass
class Index:
    value0: expr
    value1: expr

@dataclasses.dataclass
class Inv:
    value0: token
    value1: expr

@dataclasses.dataclass
class Len:
    value0: token
    value1: expr

@dataclasses.dataclass
class Neg:
    value0: token
    value1: expr

@dataclasses.dataclass
class NestedExp:
    value0: token
    value1: expr

@dataclasses.dataclass
class Nil:
    value: token

@dataclasses.dataclass
class Not:
    value0: token
    value1: expr

@dataclasses.dataclass
class Num:
    value0: token

@dataclasses.dataclass
class String:
    value0: token

@dataclasses.dataclass
class TableExpr:
    value0: table

@dataclasses.dataclass
class UnsolvedBin:
    value: list[op[expr]]

@dataclasses.dataclass
class Var:
    value: token

if typing.TYPE_CHECKING:
    expr = typing.Union[Var,UnsolvedBin,TableExpr,String,Num,Not,Nil,NestedExp,Neg,Len,Inv,Index,FuncDef,Exponent,Ellipse,CallMethod,CallFunc,Bool,Bin,Attr]
else:
    expr = (Var,UnsolvedBin,TableExpr,String,Num,Not,Nil,NestedExp,Neg,Len,Inv,Index,FuncDef,Exponent,Ellipse,CallMethod,CallFunc,Bool,Bin,Attr)

@dataclasses.dataclass
class block:
    suite: list[stmt]
    ret: maybe[stmt]

block = block

@dataclasses.dataclass
class range:
    low: expr
    high: expr
    step: maybe[expr]

range = range

@dataclasses.dataclass
class if_elseif:
    pos: token
    cond: expr
    elif_body: block

if_elseif = if_elseif

@dataclasses.dataclass
class if_else:
    pos: token
    else_body: block

if_else = if_else

@dataclasses.dataclass
class params:
    names: list[token]
    vararg: maybe[token]

params = params

File: runtests/ocaml_lua/lua_construct.ml
open Lua_require;;
open Lua_lexer;;
type ___used_t_head_90xasda
and funcname = 
| DotName of { value0 : funcname;value1 : tbnf_token }
| MethodName of { value0 : funcname;value1 : tbnf_token }
| VarName of { value : tbnf_token }
and arguments = 
| PositionalArgs of { value0 : tbnf_token;value1 : (expr) list }
| StringArg of { value : tbnf_token }
| TableArgs of { value : table }
and table = 
| TableConstructor of { value0 : tbnf_token;value1 : (table_field) list }
and table_field = 
| ElementField of { value0 : expr }
| IndexField of { pos : tbnf_token;value : expr;item : expr }
| NameField of { key : tbnf_token;value : expr }
and stmt = 
| Assignment of { value0 : bool;value1 : (expr) list;value2 : ((expr) list) maybe }
| BreakStmt of { value : tbnf_token }
| DoStmt of { value0 : tbnf_token;value1 : block }
| EmptyStmt of { value0 : tbnf_token }
| ExprStmt of { value : expr }
| ForInStmt of { value0 : tbnf_token;value1 : (tbnf_token) list;value2 : (expr) list;value3 : block }
| ForRangeStmt of { value0 : tbnf_token;value1 : tbnf_token;value2 : range;value3 : block }
| GotoStmt of { value0 : tbnf_token;value1 : tbnf_token }
| IfStmt of { value0 : tbnf_token;value1 : expr;value2 : block;value3 : (if_elseif) list;value4 : (if_else) maybe }
| LabelStmt of { value : tbnf_token }
| RepeatStmt of { value0 : tbnf_token;value1 : block;value2 : expr }
| ReturnStmt of { value0 : tbnf_token;value1 : (expr) list }
| WhileStmt of { value0 : tbnf_token;value1 : expr;value2 : block }
and expr = 
| Attr of { value0 : expr;value1 : tbnf_token }
| Bin of { op : tbnf_token;l : expr;r : expr }
| Bool of { value0 : tbnf_token;value1 : bool }
| CallFunc of { value0 : expr;value1 : arguments }
| CallMethod of { value0 : expr;value1 : tbnf_token;value2 : arguments }
| Ellipse of { value0 : tbnf_token }
| Exponent of { value0 : expr;value1 : expr }
| FuncDef of { pos : tbnf_token;is_local : bool;fname : (funcname) maybe;params : (params) maybe;body : block }
| Index of { value0 : expr;value1 : expr }
| Inv of { value0 : tbnf_token;value1 : expr }
| Len of { value0 : tbnf_token;value1 : expr }
| Neg of { value0 : tbnf_token;value1 : expr }
| NestedExp of { value0 : tbnf_token;value1 : expr }
| Nil of { value : tbnf_token }
| Not of { value0 : tbnf_token;value1 : expr }
| Num of { value0 : tbnf_token }
| String of { value0 : tbnf_token }
| TableExpr of { value0 : table }
| UnsolvedBin of { value : ((expr) op) list }
| Var of { value : tbnf_token }
and block = {
    suite : (stmt) list;
    ret : (stmt) maybe;
}
and range = {
    low : expr;
    high : expr;
    step : (expr) maybe;
}
and if_elseif = {
    pos : tbnf_token;
    cond : expr;
    elif_body : block;
}
and if_else = {
    pos : tbnf_token;
    else_body : block;
}
and params = {
    names : (tbnf_token) list;
    vararg : (tbnf_token) maybe;
}
let params (names, vararg) : params =
     { names;vararg }
let if_else (pos, else_body) : if_else =
     { pos;else_body }
let if_elseif (pos, cond, elif_body) : if_elseif =
     { pos;cond;elif_body }
let range (low, high, step) : range =
     { low;high;step }
let block (suite, ret) : block =
     { suite;ret }
let mk_Var (value) : expr =
    Var { value }
let mk_UnsolvedBin (value) : expr =
    UnsolvedBin { value }
let mk_TableExpr (value0) : expr =
    TableExpr { value0 }
let mk_String (value0) : expr =
    String { value0 }
let mk_Num (value0) : expr =
    Num { value0 }
let mk_Not (value0, value1) : expr =
    Not { value0;value1 }
let mk_Nil (value) : expr =
    Nil { value }
let mk_NestedExp (value0, value1) : expr =
    NestedExp { value0;value1 }
let mk_Neg (value0, value1) : expr =
    Neg { value0;value1 }
let mk_Len (value0, value1) : expr =
    Len { value0;value1 }
let mk_Inv (value0, value1) : expr =
    Inv { value0;value1 }
let mk_Index (value0, value1) : expr =
    Index { value0;value1 }
let mk_FuncDef (pos, is_local, fname, params, body) : expr =
    FuncDef { pos;is_local;fname;params;body }
let mk_Exponent (value0, value1) : expr =
    Exponent { value0;value1 }
let mk_Ellipse (value0) : expr =
    Ellipse { value0 }
let mk_CallMethod (value0, value1, value2) : expr =
    CallMethod { value0;value1;value2 }
let mk_CallFunc (value0, value1) : expr =
    CallFunc { value0;value1 }
let mk_Bool (value0, value1) : expr =
    Bool { value0;value1 }
let mk_Bin (op, l, r) : expr =
    Bin { op;l;r }
let mk_Attr (value0, value1) : expr =
    Attr { value0;value1 }
let mk_WhileStmt (value0, value1, value2) : stmt =
    WhileStmt { value0;value1;value2 }
let mk_ReturnStmt (value0, value1) : stmt =
    ReturnStmt { value0;value1 }
let mk_RepeatStmt (value0, value1, value2) : stmt =
    RepeatStmt { value0;value1;value2 }
let mk_LabelStmt (value) : stmt =
    LabelStmt { value }
let mk_IfStmt (value0, value1, value2, value3, value4) : stmt =
    IfStmt { value0;value1;value2;value3;value4 }
let mk_GotoStmt (value0, value1) : stmt =
    GotoStmt { value0;value1 }
let mk_ForRangeStmt (value0, value1, value2, value3) : stmt =
    ForRangeStmt { value0;value1;value2;value3 }
let mk_ForInStmt (value0, value1, value2, value3) : stmt =
    ForInStmt { value0;value1;value2;value3 }
let mk_ExprStmt (value) : stmt =
    ExprStmt { value }
let mk_EmptyStmt (value0) : stmt =
    EmptyStmt { value0 }
let mk_DoStmt (value0, value1) : stmt =
    DoStmt { value0;value1 }
let mk_BreakStmt (value) : stmt =
    BreakStmt { value }
let mk_Assignment (value0, value1, value2) : stmt =
    Assignment { value0;value1;value2 }
let mk_NameField (key, value) : table_field =
    NameField { key;value }
let mk_IndexField (pos, value, item) : table_field =
    IndexField { pos;value;item }
let mk_ElementField (value0) : table_field =
    ElementField { value0 }
let mk_TableConstructor (value0, value1) : table =
    TableConstructor { value0;value1 }
let mk_TableArgs (value) : arguments =
    TableArgs { value }
let mk_StringArg (value) : arguments =
    StringArg { value }
let mk_PositionalArgs (value0, value1) : arguments =
    PositionalArgs { value0;value1 }
let mk_VarName (value) : funcname =
    VarName { value }
let mk_MethodName (value0, value1) : funcname =
    MethodName { value0;value1 }
let mk_DotName (value0, value1) : funcname =
    DotName { value0;value1 }
